postgis/lwgeom_functions_basic.c

```diff
@@ -2064,6 +2064,69 @@ Datum ST_MakeEnvelope(PG_FUNCTION_ARGS)
 	PG_RETURN_POINTER(result);
 }
 
+
+PG_FUNCTION_INFO_V1(ST_TileEnvelope);
+Datum ST_TileEnvelope(PG_FUNCTION_ARGS)
+{
+	LWPOLY *poly;
+	GSERIALIZED *result;
+	uint32_t x, y, zoom;
+	int32_t worldTileSize;
+	double worldGeoSize, tileGeoSize;
+	double boundsWidth, boundsHeight;
+	double x1, y1, x2, y2;
+	int32_t srid = 3857;
+	GBOX *bounds;
+	GBOX bounds_3857;
+
+	POSTGIS_DEBUG(2, "ST_TileEnvelope called");
+
+	zoom = PG_GETARG_INT32(0);
+	x = PG_GETARG_INT32(1);
+	y = PG_GETARG_INT32(2);
+	if (PG_NARGS() > 3)
+		srid = PG_GETARG_INT32(3);
+	if (PG_NARGS() > 4)
+		bounds = (GBOX*)(PG_GETARG_POINTER(4));
+	else
+	{
+		gbox_init(&bounds_3857);
+		bounds_3857.ymin = bounds_3857.xmin = -20037508.3427892;
+		bounds_3857.ymax = bounds_3857.xmax =  20037508.3427892;
+		bounds = &bounds_3857;
+	}
+
+	boundsWidth  = bounds->xmax - bounds->xmin;
+	boundsHeight = bounds->ymax - bounds->ymin;
+	if (boundsWidth <= 0 || boundsHeight <= 0)
+		elog(ERROR, "%s: Geometric bounds are too small", __func__);
+
+	if (zoom >= 32 || zoom < 0)
+		elog(ERROR, "%s: Invalid tile zoom value, %d", __func__, zoom);
+
+	worldGeoSize = Max(boundsWidth, boundsHeight);
+	worldTileSize = 0x01 << (zoom > 31 ? 31 : zoom);
```

Comment:
1 << 31 is undefined behaviour for a int32

Reply:
and there's redundancy in the zoom > 31 test with the zoom >= 32 that is just above

Reply:
Testing a bit with -fsanitize=undefined, you need to have worldTileSize defined as uint32_t *and* qualify  the 0x01 constant with U, so that everything operates on uint32

Reply:
Makes sense on unsigned. I'm OK with the redundancy.